README.md
# Project StructFunc

1. Find longitudinal structural to functional connectivity associations
2. Predict functional connectivity

## Dataset structure

- All inputs (i.e. building blocks from other sources) are located in
  `inputs/`.
- All custom code is located in `code/`.


CHANGELOG.md
# Changelog

ded log

## 2021-07-13

Datalad is pretty cool. Yoda is cool.


ABCD/utils/abcd.py
import os
import pandas as pd
idx = pd.IndexSlice


INDEX = ['src_subject_id', 'eventname']
EVENTS = ['baseline_year_1_arm_1', '2_year_follow_up_y_arm_1']

FCON_TEMPLATE = 'rsfmri_c_ngd_{0}_ngd_{1}'
FCON = {
    'ad': 'auditory',
    'cgc': 'cingulo-opercular',
    'ca': 'cingulo-parietal',
    'dt': 'default',
    'dla': 'dorsal attention',
    'fo': 'fronto-parietal',
    'n': None,
    'rspltp': 'retrosplenial temporal',
    'smh': 'sensorimotor hand',
    'smm': 'sensorimotor mouth',
    'sa': 'salience',
    'vta': 'ventral attention',
    'vs': 'visual'
}

SCON_TEMPLATE = 'dmri_dtifa_fiberat_{0}'


def load_data(abcd_path, data, dropna=False, include_rec=None, exclude_n=None):
    """
    Load a longitudinally ordered ABCD dataset.
    * fcon: Gordon network correlations
    * scon: DTI atlas tract fractional anisotropy averages

    Params:
        abcd_path: ABCD dataset directory
        data: dataset to load
        dropna: drop subject if any column is NA
        include_rec: [fcon, scon] Filter by recommended inclusion?
        exclude_n: [fcon] Ignore None "network"?

    Returns:
        dataset: DataFrame indexed and sorted by (subject, event)
        extra: DataFrame of relevant supplementary data
            * fcon, scon: mean motion
    """
    if data == 'fcon':
        filename = 'abcd_betnet02.tsv'
    elif data == 'scon':
        filename = 'abcd_dti_p101.tsv'
    else:
        raise ValueError('Unknown dataset ' + data)

    raw = pd.read_csv(os.path.join(abcd_path, filename), sep='\t',
                      skiprows=[1], index_col=INDEX)
    raw = raw.loc[~raw.index.duplicated(keep='last')]

    # columns
    if data == 'fcon':
        fcon_codes = list(FCON.keys())
        if exclude_n:
            fcon_codes.remove('n')

        columns = []
        for i in range(len(fcon_codes)):
            for j in range(i+1):
                columns.append(FCON_TEMPLATE.format(fcon_codes[i], fcon_codes[j]))

        extra_columns = ['rsfmri_c_ngd_meanmotion']
    elif data == 'scon':
        columns = raw.columns.str.startswith(SCON_TEMPLATE.format(''))

        extra_columns = ['dmri_dti_meanmotion']
    else:
        columns = slice(None)

        extra_columns = []

    raw_cols = raw.loc[:, columns]

    # rows
    if include_rec:
        imgincl = pd.read_csv(os.path.join(abcd_path, 'abcd_imgincl01.tsv'), sep='\t',
                              skiprows=[1], index_col=INDEX)
        imgincl = imgincl.dropna(subset=['visit'])
        imgincl = imgincl.loc[~imgincl.index.duplicated(keep='last')]

        if data == 'fcon':
            inclusion = imgincl.loc[imgincl['imgincl_rsfmri_include'] == 1].index
        elif data == 'scon':
            inclusion = imgincl.loc[imgincl['imgincl_dmri_include'] == 1].index
        else:
            raise ValueError('No recommended inclusion for ' + data)

        included = raw_cols.loc[inclusion, :]
    else:
        included = raw_cols

    if dropna:
        included = included.dropna()

    subs_included = included.groupby(level=0).size()
    subs_long = subs_included.index[subs_included == len(EVENTS)]

    dataset = raw.loc[idx[subs_long, EVENTS], columns]
    extra = raw.loc[dataset.index, extra_columns]
    return dataset, extra


def get_scon_dict(abcd_path):
    """
    Builds and returns a dict of DTI atlas tract descriptions.

    Params:
        abcd_path: ABCD dataset directory

    Returns:
        scon_dict: dict with (code, description) for each tract
    """
    dti_labels = pd.read_csv(os.path.join(abcd_path, 'abcd_dti_p101.tsv'), sep='\t', nrows=1)
    code_start = SCON_TEMPLATE.format('')
    description_starts = ('Average fractional anisotropy within ', 'DTI atlas tract ')

    scon_labels = dti_labels.loc[0, dti_labels.columns.str.startswith(code_start)]
    codes = scon_labels.index.str.replace(code_start, '')
    descriptions = (scon_labels.str.replace(description_starts[0], '')
                    .str.replace(description_starts[1], '').values)

    scon_dict = dict(zip(codes, descriptions))
    return scon_dict


def load_covariates(path, simple_race=False):
    """
    Load ABCD covariates.

    Params:
        path: Covariates file
        simple_race: Simpler race with [White, Black, Asian, Other, Mixed].
            Other includes missing.

    Returns:
        covariates: DataFrame indexed by (subject, event)
    """
    covariates = pd.read_csv(path, index_col=INDEX)

    if simple_race:
        covariates['race'] = covariates['race.6level'].replace('AIAN/NHPI', 'Other').fillna('Other')
        covariates = covariates.drop('race.6level', axis=1)

    return covariates


code/aggregate_covariates.py
#!/usr/bin/env python
# coding: utf-8

import numpy as np
import pandas as pd
from pathlib import Path
from functools import reduce

from utils import abcd

ABCD_DIR = Path('inputs/ABCD')
NA_VALUES = [777, 999]


# INPUTS
dem_long = pd.read_csv(ABCD_DIR / 'abcd_lpds01.tsv', sep='\t',
                       skiprows=[1], index_col=abcd.INDEX)
dem_site = pd.read_csv(ABCD_DIR / 'abcd_lt01.tsv', sep='\t',
                       skiprows=[1], index_col=abcd.INDEX)
dem_base = pd.read_csv(ABCD_DIR / 'pdem02.tsv', sep='\t',
                       skiprows=[1], index_col=abcd.INDEX)
dem_acs = pd.read_csv(ABCD_DIR / 'acspsw03.tsv', sep='\t',
                      skiprows=[1], index_col=abcd.INDEX)

# start aggregate
covariates = dem_long[['interview_date', 'interview_age']].copy()

covariates['sex'] = dem_long['sex'].astype('category')
covariates['site_id'] = dem_site['site_id_l']

def combine_dems(dems, na_values=None):
    combined = reduce(
        lambda l,r: l.combine_first(r),
        [dem.replace(na_values, np.nan) for dem in dems]
    )
    return combined

# income
comb_income = combine_dems([dem_base['demo_comb_income_v2'],
                            dem_long['demo_comb_income_v2_l']],
                           na_values=NA_VALUES)

def map_income_3level(x):
    if x in np.arange(1,7):
        return '[<50K]'
    elif x in [7,8]:
        return '[>=50K & <100K]'
    elif x in [9,10]:
        return '[>=100K]'
    else:
        return np.nan

covariates['comb_income.3level'] = comb_income.apply(map_income_3level).astype('category')

# education
prnt_ed = combine_dems([dem_base['demo_prnt_ed_v2'],
                        dem_long['demo_prnt_ed_v2_l'],
                        dem_long['demo_prnt_ed_v2_2yr_l']],
                       na_values=NA_VALUES)
prtnr_ed = combine_dems([dem_base['demo_prtnr_ed_v2'],
                         dem_long['demo_prtnr_ed_v2_l'],
                         dem_long['demo_prtnr_ed_v2_2yr_l']],
                        na_values=NA_VALUES)
highest_ed = prnt_ed.fillna(-1).combine(prtnr_ed.fillna(-1), max)

def map_ed_5level(x):
    if x in np.arange(0,13):
        return '< HS Diploma'
    elif x in [13,14]:
        return 'HS Diploma/GED'
    elif x in [15,16,17,22,23]:
        return 'Some College'
    elif x == 18:
        return 'Bachelor'
    elif x in [19,20,21]:
        return 'Post Graduate Degree'
    else:
        return np.nan

covariates['highest_ed.5level'] = highest_ed.apply(map_ed_5level).astype('category')

# marital
marital = combine_dems([dem_base['demo_prnt_marital_v2'],
                        dem_long['demo_prnt_marital_v2_l']],
                       na_values=NA_VALUES)

def map_marital(x):
    if x == 1:
        return 'Yes'
    elif x in np.arange(2,7):
        return 'No'
    else:
        return np.nan

covariates['married'] = marital.apply(map_marital).astype('category')

# race
white = dem_base['demo_race_a_p___10']
black = dem_base['demo_race_a_p___11']
aian = dem_base['demo_race_a_p___12'] | dem_base['demo_race_a_p___13']
nhpi = reduce(lambda l,r: l | r,
              [dem_base['demo_race_a_p___' + str(n)] for n in range(14,18)])
asian = reduce(lambda l,r: l | r,
               [dem_base['demo_race_a_p___' + str(n)] for n in range(18,25)])
other = dem_base['demo_race_a_p___25']
mixed = white + black + asian + aian + nhpi + other > 1

race = pd.Series(np.nan, index=mixed.index, name='race.6level')
race[white == 1] = 'White'
race[black == 1] = 'Black'
race[asian == 1] = 'Asian'
race[(aian == 1) | (nhpi == 1)] = 'AIAN/NHPI'
race[other == 1] = 'Other'
race[mixed] = 'Mixed'

covariates = covariates.join(race.droplevel('eventname').astype('category'))

covariates = covariates.join(dem_base['demo_ethn_v2'].map({1:'Yes', 2:'No'}).rename('hisp')
                             .droplevel('eventname').astype('category'))

# family ID
covariates = covariates.join(
    dem_acs.xs('baseline_year_1_arm_1', level='eventname')['rel_family_id'].astype(int)
)

# OUTPUTS
covariates.to_csv('outputs/abcd_covariates.csv')


code/makeDEAPdemos.R

makeDEAPdemos<-function(datapath){

library(dplyr)  
library(data.table)
  
  acsdems<-read.delim(paste0(datapath, 'acspsw03.txt'), skip=2, header=F, sep = "\t")
  cols_acsdems<-read.delim(paste0(datapath, 'acspsw03.txt'), header=T, sep = "\t")
  colnames(acsdems)<-colnames(cols_acsdems)
  acsdems<-acsdems[,-c(which(colnames(acsdems) %in% c('collection_id','dataset_id')))]
  
basedems<-read.delim(paste0(datapath, 'pdem02.txt'), skip=2, header=F, sep = "\t")
cols_basedems<-read.delim(paste0(datapath, 'pdem02.txt'), header=T, sep = "\t")
colnames(basedems)<-colnames(cols_basedems)
basedems<-basedems[,-c(which(colnames(basedems) %in% c('collection_id','dataset_id')))]

longdems<-read.delim(paste0(datapath, 'abcd_lpds01.txt'), skip=2, header=F, sep = "\t")
cols_longdems<-read.delim(paste0(datapath, 'abcd_lpds01.txt'), header=T, sep = "\t")
colnames(longdems)<-colnames(cols_longdems)
longdems<-longdems[,-c(which(colnames(longdems) %in% c('collection_id','dataset_id')))]

lt_site<-read.delim(paste0(datapath, 'abcd_lt01.txt'), skip=2, header=F, sep = "\t")
cols_lt_site<-read.delim(paste0(datapath, 'abcd_lt01.txt'), header=T, sep = "\t")
colnames(lt_site)<-colnames(cols_lt_site)
lt_site<-lt_site[,-c(which(colnames(lt_site) %in% c('collection_id','dataset_id')))]

alldems<-join(longdems,basedems)

alldems[,'demo_prnt_ed_p']<-coalesce(alldems$demo_prnt_ed_v2,alldems$demo_prnt_ed_v2_l)
alldems[,'demo_prnt_ed_p']<-coalesce(alldems$demo_prnt_ed_p,alldems$demo_prnt_ed_v2_2yr_l)
alldems$demo_prtnr_ed_v2_l = as.integer(alldems$demo_prtnr_ed_v2_l)
alldems[,'demo_prtnr_ed_p']<-coalesce(alldems$demo_prtnr_ed_v2,alldems$demo_prtnr_ed_v2_l)
alldems[,'demo_prtnr_ed_p']<-coalesce(alldems$demo_prtnr_ed_p,alldems$demo_prtnr_ed_v2_2yr_l)

alldems[,'demo_comb_income_p']<-coalesce(alldems$demo_comb_income_v2,alldems$demo_comb_income_v2_l)

alldems[,'demo_prnt_marital_p']<-coalesce(alldems$demo_prnt_marital_v2, alldems$demo_prnt_marital_v2_l)

alldems<-join(alldems,lt_site)

deap_demos<-data.frame(alldems[,c('src_subject_id','eventname','interview_date','interview_age','sex')])

deap_demos$abcd_site = alldems$site_id_l #site_id_l is in longitudianl tracking instrument

deap_demos$sex[which(deap_demos$sex=="")]=NA
deap_demos$sex=factor( deap_demos$sex, levels= c("F","M"))

### Household income
household.income = alldems$demo_comb_income_p_all
household.income[alldems$demo_comb_income_p == "1"] = 1 # "[<50K]"
household.income[alldems$demo_comb_income_p == "2"] = 1 # "[<50K]"
household.income[alldems$demo_comb_income_p == "3"] = 1 # "[<50K]"
household.income[alldems$demo_comb_income_p == "4"] = 1 # "[<50K]"
household.income[alldems$demo_comb_income_p == "5"] = 1 # "[<50K]"
household.income[alldems$demo_comb_income_p == "6"] = 1 # "[<50K]"
household.income[alldems$demo_comb_income_p == "7"] = 2 # "[>=50K & <100K]"
household.income[alldems$demo_comb_income_p == "8"] = 2 # "[>=50K & <100K]"
household.income[alldems$demo_comb_income_p == "9"] = 3 # "[>=100K]"
household.income[alldems$demo_comb_income_p == "10"] = 3 # "[>=100K]"
household.income[alldems$demo_comb_income_p == "777"] = NA
household.income[alldems$demo_comb_income_p == "999"] = NA
household.income[household.income %in% c(NA, "999", "777")] = NA
deap_demos$household.income = factor( household.income, levels= 1:3, labels = c("[<50K]", "[>=50K & <100K]", "[>=100K]") )
#highest education: 5 different levels. These levels correspond to the numbers published by the American Community Survey (ACS). 
high.educ1 = alldems$demo_prnt_ed_p
high.educ2 = alldems$demo_prtnr_ed_p
high.educ1[which(high.educ1 == "999")] = NA
high.educ2[which(high.educ2 == "999")] = NA
high.educ1[which(high.educ1 == "777")] = NA
high.educ2[which(high.educ2 == "777")] = NA
high.educ = pmax(as.numeric(as.character(high.educ1)), as.numeric(as.character(high.educ2)), na.rm=T)
idx <- which(high.educ %in% 0:12, arr.ind = TRUE)
high.educ[idx] = 1 # "< HS Diploma"
idx <- which(high.educ %in% 13:14, arr.ind = TRUE)
high.educ[idx] = 2 # "HS Diploma/GED"
idx <- which(high.educ %in% c(15:17,22:23), arr.ind = TRUE)
high.educ[idx] = 3 # "Some College"
idx <- which(high.educ == 18, arr.ind = TRUE)
high.educ[idx] = 4 # "Bachelor"
idx <- which(high.educ %in% 19:21, arr.ind = TRUE)
high.educ[idx] = 5 # "Post Graduate Degree"
high.educ[which(high.educ == "999")]=NA
high.educ[which(high.educ == "777")]=NA
deap_demos$high.educ = factor( high.educ, levels= 1:5, labels = c("< HS Diploma","HS Diploma/GED","Some College","Bachelor","Post Graduate Degree") )
### Marrital status
married = rep(NA, length(alldems$demo_prnt_marital_p))
married[alldems$demo_prnt_marital_p == 1] = 1
married[alldems$demo_prnt_marital_p %in% 2:6] = 0
deap_demos$married = factor( married, levels= 0:1, labels = c("No", "Yes") )
#Add another variable that also includes couples that just live together. 
#married.livingtogether = rep(NA, length(alldems$demo_prnt_marital_p))
#married.livingtogether[alldems$demo_prnt_marital_p %in% c(1,6)] = 1
#married.livingtogether[alldems$demo_prnt_marital_p %in% 2:5] = 0
#deap_demos$married.or.livingtogether = factor( married.livingtogether, levels= 0:1, labels = c("no", "yes") )

#########################################

#Race of child not asked again longitudinally? Repeating this from baseline
race_cols = c("demo_ethn_v2", "demo_race_a_p___10", "demo_race_a_p___11","demo_race_a_p___12", "demo_race_a_p___13",
              "demo_race_a_p___14", "demo_race_a_p___15", "demo_race_a_p___16", "demo_race_a_p___17", 
              "demo_race_a_p___18", "demo_race_a_p___19", "demo_race_a_p___20", "demo_race_a_p___21", "demo_race_a_p___22","demo_race_a_p___23",
              "demo_race_a_p___24", "demo_race_a_p___25",
              "demo_race_a_p___77", "demo_race_a_p___99")
racedf<-basedems[,match(c('src_subject_id','eventname',race_cols), names(basedems))]

raceind<-which(colnames(alldems) %in% race_cols)
alldems<-alldems[,-c(raceind)]

alldems<-merge(alldems, racedf, by.x = 'src_subject_id', by.y = 'src_subject_id')

raceind<-which(colnames(alldems) %in% race_cols)
dat = data.table(alldems[,raceind])

# White
dat[, white:= (demo_race_a_p___10 == 1)*1 ]

# Black
dat[, black:= (demo_race_a_p___11 == 1)*1 ]

# Asian
dat[, asian:= 0]
dat[ (demo_race_a_p___18 == 1 | demo_race_a_p___19 == 1 | demo_race_a_p___20 == 1 |
        demo_race_a_p___21 == 1 | demo_race_a_p___22 == 1 | demo_race_a_p___23 == 1 |
        demo_race_a_p___24==1), asian:= 1 ]

# AIAN: American Indian and Alaska Native
dat[, aian:= 0]
dat[ (demo_race_a_p___12 == 1 | demo_race_a_p___13 == 1), aian:=1 ]

#NHPI: Native Hawaiian and Other Pacific
dat[, nhpi:= 0]
dat[ demo_race_a_p___14 == 1 | demo_race_a_p___15 == 1 | demo_race_a_p___16 == 1 |
       demo_race_a_p___17 == 1, nhpi:= 1 ]

# Other
dat[, other:= 0 ]
dat[ demo_race_a_p___25 == 1, other:= 1 ]

# Mixed
dat[, mixed:= (white + black + asian + aian + nhpi + other)]
dat[, table(mixed, useNA = "if")]
dat[ mixed <= 1, mixed:= 0]
dat[ mixed > 1, mixed:= 1]
dat[, table(mixed, useNA = "if")]

# Race 4 level
dat[ white == 1, race.4level:= 1]
dat[ black == 1,race.4level:= 2]
dat[ asian == 1,race.4level:= 3]
dat[ aian == 1,race.4level:= 4]
dat[ nhpi == 1,race.4level:= 4]
dat[ other == 1,race.4level:= 4]
dat[ mixed == 1,race.4level:= 4]
dat[, table(race.4level, useNA = "if") ]

dat$race.4level<- factor(dat$race.4level,
                         levels = 1:4,
                         labels = c("White","Black","Asian","Other/Mixed"))

dat$race.eth[dat$race.4level==1] = "White"
dat$race.eth[dat$race.4level==2] = "Black"
dat$race.eth[dat$race.4level==3] ="Asian"
dat$race.eth[dat$race.4level==4] = "Other/Mixed"
dat[, table(race.4level, useNA = "if") ]

# Race 6 level
dat[white==1,race.6level:=1]
dat[black==1,race.6level:=2]
dat[asian==1,race.6level:=3]
dat[aian==1,race.6level:=4]
dat[nhpi==1,race.6level:=4]
dat[other==1,race.6level:=5]
dat[mixed==1,race.6level:=6]
dat[, table(race.6level,useNA="if") ]

dat$race.6level<- factor(dat$race.6level,
                         levels=1:6,
                         labels= c("White","Black","Asian","AIAN/NHPI","Other","Mixed"))
dat$race.eth[dat$race.6level==1] ="White"
dat$race.eth[dat$race.6level==2] ="Black"
dat$race.eth[dat$race.6level==3] ="Asian"
dat$race.eth[dat$race.6level==4] ="AIAN/NHPI"
dat$race.eth[dat$race.6level==5] ="Other"
dat$race.eth[dat$race.6level==6] ="Mixed"
dat[, table(race.6level,useNA="if") ]

# Hispanic
dat$hisp=NA;
indx.1=which(dat$demo_ethn_v2==1)
indx.0=which(dat$demo_ethn_v2==2)
dat$hisp[indx.1]=1;
dat$hisp[indx.0]=0;


dat$hisp<- factor(dat$hisp,
                  levels=0:1,
                  labels= c("No","Yes"))

dat$src_subject_id<-alldems$src_subject_id
dat$eventname<-alldems$eventname.x

racetmp<-data.frame(dat[,c('src_subject_id','eventname','race.4level','race.6level','hisp')])

deap_demos<-join(deap_demos,racetmp)

############################################ 

#Replicating family ID over all time points

alldems<-merge(alldems, acsdems[acsdems$eventname=='baseline_year_1_arm_1',c('src_subject_id','rel_family_id')])
alldems$eventname<-alldems$eventname.x

famtmp<-data.frame(alldems[,c('src_subject_id','eventname','rel_family_id')])

deap_demos<-join(deap_demos,famtmp)


return(deap_demos)
}
